File: src/main/java/komu/blunt/core/CoreLambdaExpression.java
package komu.blunt.core;

import komu.blunt.asm.*;

import static com.google.common.base.Preconditions.checkArgument;
import static com.google.common.base.Preconditions.checkNotNull;

public final class CoreLambdaExpression extends CoreExpression {

    private final int envSize;
    private final CoreExpression body;

    public CoreLambdaExpression(int envSize, CoreExpression body) {
        checkArgument(envSize >= 0);

        this.envSize = envSize;
        this.body = checkNotNull(body);
    }

    @Override
    public void assemble(Assembler asm, Instructions instructions, Register target, Linkage linkage) {
        // TODO: place the lambda in a new code section
        Label lambda = asm.newLabel("lambda");
        Label afterLambda = asm.newLabel("after-lambda");

        instructions.loadLambda(target, lambda);
        if (linkage == Linkage.NEXT) {
            instructions.jump(afterLambda);
        } else {
            instructions.finishWithLinkage(linkage);
        }

        instructions.label(lambda);
        instructions.createEnvironment(envSize);
        body.assemble(asm, instructions, Register.VAL, Linkage.RETURN);
        instructions.label(afterLambda);
    }
}

File: src/main/kotlin/komu/blunt/core/CoreLambdaExpression.kt
package komu.blunt.core

import komu.blunt.asm.*

class CoreLambdaExpression(private val envSize: Int, private val body: CoreExpression) : CoreExpression() {

    init {
        require(envSize >= 0)
    }

    override fun assemble(asm: Assembler, target: Register, linkage: Linkage) =
        instructions {
            // TODO: place the lambda in a new code section
            val lambda = asm.newLabel("lambda")
            val afterLambda = asm.newLabel("after-lambda")

            loadLambda(target, lambda)
            if (linkage == Linkage.Next)
                jump(afterLambda)
            else
                finishWithLinkage(linkage)

            label(lambda)
            createEnvironment(envSize)
            instructionsOf(body.assemble(asm, Register.VAL, Linkage.Return))
            label(afterLambda)
        }

    override fun simplify() = CoreLambdaExpression(envSize, body.simplify())

    override fun toString() = "(lambda $envSize $body)"
}
